airbyte-cdk/bulk/toolkits/load-iceberg-parquet/src/main/kotlin/io/airbyte/cdk/load/toolkits/iceberg/parquet/io/IcebergTableCleaner.kt

```diff
@@ -54,16 +56,42 @@ class IcebergTableCleaner(private val icebergUtil: IcebergUtil) {
                 }
                 .toSet()
 
-        table.newScan().planFiles().use { tasks ->
-            tasks
-                .filter { task -> genIdsToDelete.any { id -> task.file().location().contains(id) } }
-                .forEach { task ->
-                    table
-                        .newDelete()
-                        .toBranch(stagingBranchName)
-                        .deleteFile(task.file().location())
-                        .commit()
-                }
+        if (stream.isSingleGenerationTruncate()) {
+            // WARNING: This approach only works if users do not run compaction while the truncate
+            // refresh sync is running.If compaction occurs during the sync, files from the current
+            // generation may be renamed and lose their generation ID suffix, causing them to be
+            // incorrectly deleted and resulting in data loss.
+            // -------
+            // TODO: We need to eventually be able to handle to detect when a compaction was run.
+            // If it was run during the truncate refresh run then we should remove that snapshot
+            // or more like cherry pick the "valid" ones
+            val currentGenId = icebergUtil.constructGenerationIdSuffix(stream.minimumGenerationId)
+
+            table.newScan().planFiles().use { tasks ->
+                tasks
+                    .filter { task -> !task.file().location().contains(currentGenId) }
+                    .forEach { task ->
+                        table
+                            .newDelete()
+                            .toBranch(stagingBranchName)
+                            .deleteFile(task.file().location())
+                            .commit()
+                    }
+            }
+        } else {
+            table.newScan().planFiles().use { tasks ->
+                tasks
+                    .filter { task ->
+                        genIdsToDelete.any { id -> task.file().location().contains(id) }
+                    }
+                    .forEach { task ->
+                        table
+                            .newDelete()
+                            .toBranch(stagingBranchName)
+                            .deleteFile(task.file().location())
+                            .commit()
+                    }
+            }
         }
```

Comment:
not sure I understand this change - wouldn't we want to do the blind delete in both cases? I.e. we could just leave the old code alone, and add the comment that it causes data loss if the user runs compaction